cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll

```diff
@@ -0,0 +1,88 @@
+/**
+ * Provides implementation classes modeling `sscanf`, `fscanf` and various similar
+ * functions. See `semmle.code.cpp.models.Models` for usage information.
+ */
+
+import semmle.code.cpp.Function
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.Taint
+import semmle.code.cpp.models.interfaces.Alias
+import semmle.code.cpp.models.interfaces.SideEffect
+
+/**
+ * The standard function `sscanf`, `fscanf` and its assorted variants
+ */
+private class Sscanf extends ArrayFunction, TaintFunction, AliasFunction, SideEffectFunction {
+  Sscanf() {
+    this.hasGlobalOrStdName([
+        "sscanf", // sscanf(src_stream, format, args...)
+        "swscanf", // swscanf(src, format, args...)
+        "fscanf", // fscanf(src_stream, format, args...)
+        "fwscanf" // fwscanf(src_stream, format, args...)
+      ]) or
+    this.hasGlobalName([
+        "_sscanf_l", // _sscanf_l(src, format, locale, args...)
+        "_swscanf_l", // _swscanf_l(src, format, locale, args...)
+        "_snscanf", // _snscanf(src, length, format, args...)
+        "_snscanf_l", // _snscanf_l(src, length, format, locale, args...)
+        "_snwscanf", // _snwscanf(src, length, format, args...)
+        "_snwscanf_l", // _snwscanf_l(src, length, format, locale, args...)
+        "_fscanf_l", // _fscanf_l(src_stream, format, locale, args...)
+        "_fwscanf_l" // _fwscanf_l(src_stream, format, locale, args...)
+      ])
+  }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) {
+    bufParam = [0, getFormatPosition()]
```

Comment:
Is this true for all variations of the function?
I'm thinking especially about `fscanf` that takes a `FILE` pointer at position 0.

Reply:
Good point! I think the `0`'th parameter should only be included for the `sscanf` functions.